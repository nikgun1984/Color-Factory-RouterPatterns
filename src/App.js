import './App.css';
import {BrowserRouter as Router,Route,Redirect, Switch} from 'react-router-dom';
import {useState} from 'react';
import { v4 as uuid } from "uuid";
import { Colors } from "./Colors";
import { Color } from "./Color";
import { ColorForm } from "./ColorForm";
import NavBar from "./NavBar";

function App() {
  const defaultColors = [
		{ text: "red", color: "#FF0000",id: uuid() },
		{text: "green",color:"#11FF00", id:uuid()},
		{text:"blue",color:"#0037FF",id:uuid()}
	];
  const [colors, setColors] = useState(defaultColors);
	const addColor = (color) => {
		const newColor = {...color,id:uuid()};
		setColors((colors) => [newColor,...colors]);
	};
  return (
		<div className="App">
			<Router>
				<Switch>
					<Route exact path="/colors/new">
						<ColorForm colors={colors} addColor={addColor}/>
					</Route>
					<Route exact path="/colors/:color">
						<Color />
					</Route>
					<Route exact path="/colors/">
						<NavBar />
						<Colors colors={colors}/>
					</Route>
					<Redirect to="/colors" />
				</Switch>
			</Router>
		</div>
	);
}

export default App;
