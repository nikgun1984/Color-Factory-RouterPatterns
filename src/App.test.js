import { render, fireEvent } from "@testing-library/react";
import App from "./App";
import { MemoryRouter } from "react-router-dom";

//smoke test
it("mounts without crashing", function () {
	const { getByText } = render(
		<MemoryRouter initialEntries={["/colors/"]}>
			<App />
		</MemoryRouter>
	);
	const redLink = getByText("red");
	expect(redLink).toBeInTheDocument();
});

//snapshot tests
it("matches snapshot", function () {
	const { asFragment } = render(<App />);
	expect(asFragment()).toMatchSnapshot();
});

it("navbar links like red", function () {
	const { getByText } = render(
		<MemoryRouter initialEntries={["/colors/"]}>
			<App />
		</MemoryRouter>
	);

	const link = getByText("red");
	fireEvent.click(link);
	expect(getByText("ISN'T IT BEAUTIFUL")).toBeInTheDocument();

	const backLink = getByText("GO BACK HOME");
	fireEvent.click(backLink);
	expect(getByText("Welcome to the color factory.")).toBeInTheDocument();
});

it("checks if form submits", function () {
	const { getByText, getByLabelText, queryByText } = render(
		<MemoryRouter initialEntries={["/colors/new"]}>
			<App />
		</MemoryRouter>
	);
	const addColorLink = getByText("Add a Color");
	expect(addColorLink).toBeInTheDocument();

	fireEvent.click(addColorLink);

	// no items yet
	expect(queryByText("pink")).not.toBeInTheDocument();

	//fill out the form
	const textInput = getByLabelText("Name Color");
	const colorInput = getByLabelText("Color");
	const submitBtn = queryByText("Add this Color");

	fireEvent.change(textInput, { target: { value: "pink" } });
	fireEvent.change(colorInput, { target: { value: "#FF00FF" } });
	fireEvent.click(submitBtn);

	expect(queryByText("pink")).toBeInTheDocument();
});
