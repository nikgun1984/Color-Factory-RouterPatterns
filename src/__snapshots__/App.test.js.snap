// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="App-header"
    >
      <h3>
        Welcome to the color factory.
      </h3>
      <h1>
        <a
          href="/colors/new"
        >
          Add a Color
        </a>
      </h1>
    </nav>
    <div>
      <a
        href="/colors/red"
      >
        <p>
          red
        </p>
      </a>
      <a
        href="/colors/green"
      >
        <p>
          green
        </p>
      </a>
      <a
        href="/colors/blue"
      >
        <p>
          blue
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
